app/services/fees/subscription_service.rb

```diff
@@ -29,21 +50,39 @@ def create
         taxes_amount_cents: 0,
         unit_amount_cents: new_amount_cents,
       )
-      new_fee.precise_unit_amount = new_fee.unit_amount.to_f
-      new_fee.save!
 
-      result.fee = new_fee
-      result
-    rescue ActiveRecord::RecordInvalid => e
-      result.record_validation_failure!(record: e.record)
-    end
+      return base_fee unless invoice.draft?
+      return base_fee unless adjusted_fee
 
-    private
+      units = adjusted_fee.units
+      unit_amount_cents = adjusted_fee.unit_amount_cents.round
+      amount_cents = adjusted_fee.adjusted_units? ? (units * new_amount_cents) : (units * unit_amount_cents)
+      Fee.new(
+        invoice:,
+        subscription:,
+        amount_cents: amount_cents.round,
+        amount_currency: plan.amount_currency,
+        fee_type: :subscription,
+        invoiceable_type: 'Subscription',
+        invoiceable: subscription,
+        units:,
+        properties: boundaries.to_h,
+        payment_status: :pending,
+        taxes_amount_cents: 0,
+        unit_amount_cents: adjusted_fee.adjusted_units? ? new_amount_cents : unit_amount_cents,
+        invoice_display_name: adjusted_fee.invoice_display_name,
+      )
+    end
 
-    attr_reader :invoice, :subscription, :boundaries
+    def adjusted_fee
+      return @adjusted_fee if defined? @adjusted_fee
 
-    delegate :customer, to: :invoice
-    delegate :previous_subscription, :plan, to: :subscription
+      @adjusted_fee = AdjustedFee
+        .where(invoice:, subscription:, fee_type: :subscription)
+        .where("(properties->>'from_datetime')::date = ?", boundaries.from_datetime.to_date)
+        .where("(properties->>'to_datetime')::date = ?", boundaries.to_datetime.to_date)
```

Comment:
I think it would also work with dates since same timestamp as original one is used to refresh invoice and calculate boundaries (so the date should also match the one in the DB) but refactored anyway and converted the query to work with datetimes